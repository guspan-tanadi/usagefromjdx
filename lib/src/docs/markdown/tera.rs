use itertools::Itertools;
use once_cell::sync::Lazy;
use std::collections::HashMap;
use tera::Tera;

pub(crate) static TERA: Lazy<Tera> = Lazy::new(|| {
    let mut tera = Tera::default();

    #[rustfmt::skip]
    tera.add_raw_templates([
        ("arg_template.md.tera", include_str!("templates/arg_template.md.tera")),
        ("cmd_template.md.tera", include_str!("templates/cmd_template.md.tera")),
        ("flag_template.md.tera", include_str!("templates/flag_template.md.tera")),
        ("spec_template.md.tera", include_str!("templates/spec_template.md.tera")),
        ("index_template.md.tera", include_str!("templates/index_template.md.tera")),
    ]).unwrap();

    tera.register_filter(
        "repeat",
        move |value: &tera::Value, args: &HashMap<String, tera::Value>| {
            let value = value.as_str().unwrap();
            let count = args.get("count").unwrap().as_u64().unwrap();
            Ok(value.repeat(count as usize).into())
        },
    );

    tera.register_filter(
        "escape_md",
        move |value: &tera::Value, _: &HashMap<String, tera::Value>| {
            let value = value.as_str().unwrap();
            let value = value
                .lines()
                .map(|line| {
                    if line.starts_with("    ") {
                        return line.to_string();
                    }
                    // replace '<' with '&lt;' but not inside code blocks
                    xx::regex!(r"(`[^`]*`)|(<)")
                        .replace_all(line, |caps: &regex::Captures| {
                            if caps.get(1).is_some() {
                                caps.get(1).unwrap().as_str().to_string()
                            } else {
                                "&lt;".to_string()
                            }
                        })
                        .to_string()
                })
                .join("\n");
            Ok(value.into())
        },
    );

    tera
});
