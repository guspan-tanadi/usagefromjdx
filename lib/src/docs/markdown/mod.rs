use crate::error::UsageErr;
use tera::{Context, Tera};

const SPEC_TEMPLATE: &str = include_str!("cmd_template.tera");

impl crate::spec::Spec {
    pub fn render_markdown(&self) -> Result<String, UsageErr> {
        let mut ctx = Context::new();
        ctx.insert("header", "#");
        ctx.insert("bin", &self.bin);
        ctx.insert("cmd", &self.cmd);
        let out = Tera::one_off(SPEC_TEMPLATE, &ctx, false)?;
        Ok(out)
    }
}

#[cfg(test)]
mod tests {
    use crate::Spec;
    use insta::assert_snapshot;

    #[test]
    fn test_render_markdown() {
        let spec: Spec = r#"
        bin "mycli"
        arg "arg1" help="arg1 description"
        arg "arg2" help="arg2 description" default="default value" {
            choices "choice1" "choice2" "choice3"
        }
        arg "arg3" help="arg3 description" required=true long_help="arg3 long description"
        arg "argrest" var=true
        
        flag "--flag1" help="flag1 description"
        flag "--flag2" help="flag2 description" long_help="flag2 long description"
        flag "--flag3" help="flag3 description" negate="--no-flag3"
        "#
        .parse()
        .unwrap();
        assert_snapshot!(spec.render_markdown().unwrap());
    }
}
