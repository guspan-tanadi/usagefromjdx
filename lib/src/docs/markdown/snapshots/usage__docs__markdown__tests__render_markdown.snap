---
source: lib/src/docs/markdown/mod.rs
expression: spec.render_markdown().unwrap()
---
# `mycli`

## Arg `<arg1>`

(required) arg1 description

## Arg `<arg2>`

(required) arg2 description

## Arg `<arg3>`

(required) arg3 long description

## Arg `<argrest>...`

(required) 

## Flag `--flag1`

flag1 description

## Flag `--flag2`

flag2 long description

## Flag `--flag3`

flag3 description
